homeassistant/components/aurora_abb_powerone/__init__.py

```diff
@@ -1 +1,73 @@
 """The Aurora ABB Powerone PV inverter sensor integration."""
+
+# Reference info:
+# https://s1.solacity.com/docs/PVI-3.0-3.6-4.2-OUTD-US%20Manual.pdf
+# http://www.drhack.it/images/PDF/AuroraCommunicationProtocol_4_2.pdf
+
+import asyncio
+from collections import defaultdict
+import logging
+
+from aurorapy.client import AuroraSerialClient
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ADDRESS, CONF_PORT
+from homeassistant.core import HomeAssistant
+
+from .const import ATTR_SERIAL_NUMBER, DOMAIN
+
+PLATFORMS = ["sensor"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Aurora ABB PowerOne component."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Aurora ABB PowerOne from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    _LOGGER.debug("async_setup_entry got user config data=%s", entry.data)
+
+    comport = entry.data[CONF_PORT]
+    address = entry.data[CONF_ADDRESS]
+    client = AuroraSerialClient(address, comport, parity="N", timeout=1)
+    all_device_params = [
+        {"type": "sensor", "parameter": "instantaneouspower", "name": "Power Output"},
+        {"type": "sensor", "parameter": "temperature", "name": "Temperature"},
+    ]
+
+    entry_data = hass.data[DOMAIN][entry.entry_id] = {
+        "client": client,
+        "devices": defaultdict(list),
+    }
+    serialnum = entry.data[ATTR_SERIAL_NUMBER]
+    entry_data["devices"][serialnum] = defaultdict(list)
+    for device_params in all_device_params:
+        entry_data["devices"][serialnum][device_params["type"]].append(device_params)
```

Comment:
This makes little sense. 

Why not define this right when you define the data structure like this:

```python
entry_data = hass.data[DOMAIN]entry.entry_id] = {
  "devices": {
    entry.data[ATTR_SERIAL_NUMBER]: [
        {"type": "sensor", "parameter": "instantaneouspower", "name": "Power Output"},
        {"type": "sensor", "parameter": "temperature", "name": "Temperature"},
    ]
  }
}
```

Reply:
I also notice that you only ever extract devices from this dict using the serial number. Since each config entry is a single serial number, and you instantiate a dict per config entry, you don't need to use serial number at all ?

Reply:
Thanks for the review!  Probably I have misunderstood something. I thought it was preferred to assign a unique id, and this was attempted [before](https://github.com/home-assistant/core/pull/31649), but rejected.

The serial number is used for the unique ID. It is possible to have multiple inverters on the same com port (but rare).

Reply:
I agree with the suggestions.

BTW I got a lot of the design patterns from the [Tesla integration](https://github.com/home-assistant/core/blob/e47d8da3b4a1b6bcd7dff11a8d5c1b603bc67a41/homeassistant/components/tesla/__init__.py#L174-L198), so the same comments about defaultdict and how to initialise it might apply there too.

Reply:
Fixed. No longer using entry.data[ATTR_SERIAL_NUMBER]